ci(contracts): enable latest artifact fallback logic

aliersh opened the pull request:
### Changes

Enables the artifact fallback mechanism introduced in [#18291](
https://github.com/ethereum-optimism/optimism/pull/18291) to optimize PR contract test builds:

- `pull-artifacts.sh`: Added `--fallback-to-latest` flag that tries exact checksum first, then falls back to `artifacts-v1-latest.tar.{gz,zst}` uploaded by develop
- `.circleci/config.yml`: Enabled fallback for `contracts-bedrock-tests`, `contracts-bedrock-coverage`, and `contracts-bedrock-tests-upgrade` jobs
- Extracted `download_and_extract()` helper function to eliminate duplication
- Added `--exclude='*../*'` to tar commands to prevent path traversal attacks

### Reason

This completes the two-phase rollout started in [#18291](https://github.com/ethereum-optimism/optimism/pull/18291). PR builds currently compile all contracts from scratch because exact checksum-based artifacts don't exist for new code. With this fallback now enabled, PRs download the most recent develop artifacts (uploaded by #18291) and only recompile changed contracts, reducing CI build time through incremental compilation instead of full rebuilds.

aliersh committed:
ci(contracts): enable latest artifact fallback logic

- add --fallback-to-latest flag to pull-artifacts.sh
- enable fallback forcontracts tests, coverage, and upgrade jobs in ci config
- extract download_and_extract helper function in pull-artifacts.sh
- add path traversal protection with --exclude='*../*' flag to tar commands

aliersh commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -16,11 +16,45 @@ echoerr() {
+  echoerr "> Done."
+
+  echoerr "> Extracting artifacts..."
+  if [[ "$archive_name" == *.tar.zst ]]; then
+    zstd -dc "$archive_name" | tar -xf - --exclude='*../*'
+  else
+    tar -xzvf "$archive_name" --exclude='*../*'
+  fi
```
The `--exclude='*../*'` flag blocks path traversal attacks by rejecting any path containing `../`. The suggested `--strip-components=1` would break the build system by removing the required top-level directories (`artifacts/`, `forge-artifacts/`, `cache/`). Since these archives are extracted from a trusted source (a controlled GCS bucket), the current protection is sufficient.

aliersh committed:
test: trigger artifact fallback by modifying test file

maurelian commented on packages/contracts-bedrock/test/L2/OperatorFeeVault.t.sol:
```diff
@@ -33,3 +33,5 @@ contract OperatorFeeVault_Version_Test is CommonTest {
     }
 }
+
+// Test artifact fallback logic
```
Is this meant to be here?

aliersh replied:
It was the change that I made to trigger the fallback in the 2nd commit. I'll delete it later

maurelian commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -16,11 +16,45 @@ echoerr() {
+  if [[ "$archive_name" == *.tar.zst ]]; then
+    zstd -dc "$archive_name" | tar -xf - --exclude='*../*'
+  else
+    tar -xzvf "$archive_name" --exclude='*../*'
```
do we actually need to handle two different compression types?

aliersh replied:
Actually, that's outside the scope of this PR, since I only extracted the existing logic to avoid duplicating code.

The change came from #18237

maurelian replied:
yeah, I see that we have other code for both types already.

maurelian commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -41,57 +75,43 @@ echoerr "> Checking for existing artifacts..."
-  echoerr "> Found .tar.gz artifacts (zstd not available)."
+  # Try latest fallback if enabled
+  if [ "$USE_LATEST_FALLBACK" = true ]; then
+    echoerr "> Exact checksum not found, trying latest artifacts..."
```
what I'm a bit confused about is when we decide to rebuild the artifacts if they actually have changed. Is there a situation where the checksum is not found, but we don't need to rebuild?

aliersh replied:
The fallback doesn't determine whether to rebuild, just provides a starting point (artifacts + cache from develop) so forge can do incremental compilation instead of rebuilding everything from scratch.

Without fallback: checksum doesn't match → no artifacts → full rebuild 
With fallback: checksum doesn't match → use latest artifacts → incremental rebuild

aliersh replied:
I don't know if that answers your question or if you are thinking of a specific scenario 🤔

maurelian replied:
oh yeah, I see what you mean; even if there are some changes most contracts will not need to be rebuilt.

This is nice.

aliersh committed:
fix(scripts): address bots security concerns in pull-artifacts.sh

- add secure curl flags (--fail, --location, --connect-timeout, --max-time, --tlsv1.2) to prevent hangs and enforce TLS
- update path traversal exclusion pattern from '*../*' to '*..*' to block all paths containing '..' anywhere
- revert test comment change used for CI testing

aliersh commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -16,11 +16,45 @@ echoerr() {
+  if [[ "$archive_name" == *.tar.zst ]]; then
+    zstd -dc "$archive_name" | tar -xf - --exclude='*../*'
+  else
+    tar -xzvf "$archive_name" --exclude='*../*'
```
fixed in https://github.com/ethereum-optimism/optimism/pull/18310/commits/9421735fd976bfa5f904c317d874e634f9aa2e87

aliersh commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -16,11 +16,45 @@ echoerr() {
+  local archive_name=$1
+
+  echoerr "> Downloading..."
+  curl -o "$archive_name" "https://storage.googleapis.com/oplabs-contract-artifacts/$archive_name"
```
fixed in https://github.com/ethereum-optimism/optimism/pull/18310/commits/9421735fd976bfa5f904c317d874e634f9aa2e87

aliersh commented on packages/contracts-bedrock/scripts/ops/pull-artifacts.sh:
```diff
@@ -16,11 +16,45 @@ echoerr() {
+  local archive_name=$1
+
+  echoerr "> Downloading..."
+  curl -o "$archive_name" "https://storage.googleapis.com/oplabs-contract-artifacts/$archive_name"
```
fixed in https://github.com/ethereum-optimism/optimism/pull/18310/commits/9421735fd976bfa5f904c317d874e634f9aa2e87
